prov/verbs/src/fi_verbs.h

```diff
@@ -957,20 +969,65 @@ vrb_send_iov_flags(struct vrb_ep *ep, struct ibv_send_wr *wr,
 		      uint64_t flags)
 {
 	size_t len = 0;
-
-	if (!desc)
-		vrb_set_sge_iov_inline(wr->sg_list, iov, count, len);
-	else
-		vrb_set_sge_iov_count_len(wr->sg_list, iov, count, desc, len);
+	enum fi_hmem_iface iface = FI_HMEM_SYSTEM;
+	uint64_t device = 0;
+	void *bounce_buf;
+	void *send_desc = NULL;
+	ssize_t ret;
+
+	if (ep->hmem_enabled) {
+		if (!desc) {
+			vrb_set_sge_iov_inline(wr->sg_list, iov, count, len);
+		} else {
+			vrb_set_sge_iov_count_len(wr->sg_list, iov, count, desc,
+						  len);
+			iface = ((struct vrb_mem_desc *) desc[0])->iface;
+			device = ((struct vrb_mem_desc *) desc[0])->device;
+			send_desc = desc[0];
+		}
+
+		wr->send_flags = VERBS_INJECT_FLAGS(ep, len, flags, send_desc);
+
+		if (wr->send_flags & IBV_SEND_INLINE &&
+		    iface != FI_HMEM_SYSTEM) {
+			bounce_buf = alloca(len);
```

Comment:
Are there any guarantees that the send buffer is in custody to the network device at the time we exit the function? Alloca allocates stack space and there is no guarantee that the data in the buffer location is valid after the function exits if the request is queued for any reason.

Reply:
If the data is not in the custody of the network device at the end of the ibv_post_send(), seems the IB network stack implementation of IBV_SEND_INLINE would be broken.

Reply:
Nevermind, I went digging into vrb_post_send. There is no chance it'll be queued, so I'm fine with the use of alloca here.